keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,750 @@
+
+    Golang types <=> OpenAPI <=> Protobuf.
+
+Each of these is to be restricted to lead to a canonical mapping.
+
+## Proposal
+
+We propose to define a subset of OpenAPI v3 called _structural schema_ following this guiding principle:
+
+**Structural Completeness**: all types and the possible object fields within a JSON object must be apparent from a structural schema without understanding the logical junctors `anyOf`, `allOf`, `oneOf` and `not`.
+
+Example:
+```yaml
+properties:
+  foo:
+    type: string
+  bar:
+    minimum: 42
+anyOf:
+- properties:
+    foo:
+      pattern: abc
+    bar:
+      type: string
+- properties:
+    bar:
+      type: integer
+      minimum: 1
+      maximum: 0
+```
+Here, `.foo` is specified as string even without looking at the `anyOf` construct, but `.bar`'s type is unclear without deeper semantic knowledge of `anyOf` and of minimum and maximum.
+
+We define a _structural schema_ by (possibly hypothetically, the actually implementation might look differently) splitting our existing `JSONSchemaProps` type into the following, while adding vendor extension fields for:
+* `x-kubernetes-preserve-unknown-fields: true` – recursive version of `additionalProperties: true`, influencing pruning to not prune unknown fields.
+* `x-kubernetes-embedded-resource: true` – marks a RawExtension field to embed a complete `runtime.Object` like object with `TypeMeta` and `ObjectMeta`.
+* `x-kubernetes-union-fields: [<string>...]`, `x-kubernetes-union-field-values: {<string>: <string>...}` and `x-kubernetes-union-discriminator: <string>` to implement [KEP: Union Types](https://github.com/kubernetes/enhancements/pull/926)
+* `x-kubernetes-int-or-string: true` – marks a `IntOrString` polymorphic type.
+
+```go
+type StructuralSchema struct {
+	Items                *StructuralSchemaOrArray
+	Properties           map[string]ValueValidation
+
+	Generic
+	Extensions
+	
+	*ValueValidation
+}
+
+type Extensions struct {
+	// x-kubernetes-union-fields lists the mutual exclusive field 
+	// names. This can only be set if x-kubernetes-union-discriminator 
+	// is empty.
+	XUnionFields           []string
+	// x-kubernetes-union-field-values lists the discriminator values 
+	// for mutual exclusive field names. This can only be set if
+	// x-kubernetes-union-discriminator is non-empty.
+        XUnionFieldValues      map[string]string
+	// x-kubernetes-union-discriminator is a field name used as the 
+	// union discrimator with the values defined in 
+	// x-kubernetes-union-field-values
+        XDiscriminator         map[string]string
+        
+        // x-kubernetes-preserve-unknown-fields stops the API server 
+        // decoding step from pruning fields which are not specified 
+        // in the validation schema. This affects fields recursively,
+        // but switches back to normal pruning behaviour if nested 
+        // properties are specified in the schema.
```

Comment:
> This affects fields recursively, but switches back to normal pruning behaviour if nested properties are specified in the schema.

Is there a clearer way to say this? Maybe something like:
> `x-kubernetes-preserve-unknown-fields: true` preserves unknown fields for this object (and any children of those unknown fields)

Thinking through this again, isn't that what `additionalProperties: true` means? "There may be unknown properties, some of which may be objects with arbitrary child fields"?

Reply:
Clarified during offline Slack discussion.